ksqldb-rest-model/src/main/java/io/confluent/ksql/rest/entity/StreamedRow.java

```diff
@@ -218,21 +368,80 @@ public boolean equals(final Object o) {
       }
       final Header header = (Header) o;
       return Objects.equals(queryId, header.queryId)
-          && Objects.equals(schema, header.schema);
+          && Objects.equals(keySchema, header.keySchema)
+          && Objects.equals(columnsSchema, header.columnsSchema);
     }
 
     @Override
     public int hashCode() {
-      return Objects.hash(queryId, schema);
+      return Objects.hash(queryId, keySchema, columnsSchema);
+    }
+  }
+
+  @JsonInclude(Include.NON_EMPTY)
+  @JsonIgnoreProperties(ignoreUnknown = true)
+  public static final class DataRow extends BaseRow {
+
+    @EffectivelyImmutable
+    private final Optional<List<?>> key;
+    @EffectivelyImmutable
+    private final Optional<List<?>> columns;
+
+    public static DataRow row(
+        final Optional<List<?>> key,
+        final List<?> columns
+    ) {
+      return new DataRow(key, Optional.of(columns));
+    }
+
+    public static DataRow tombstone(
+        final List<?> key
+    ) {
+      return new DataRow(Optional.of(key), Optional.empty());
+    }
+
+
+    public Optional<List<?>> getKey() {
+      return key;
+    }
+
+    public Optional<List<?>> getColumns() {
+      return columns;
+    }
+
+    public Optional<Boolean> getTombstone() {
```

Comment:
Why not directly return boolean?

Reply:
This is about getting the JSON right. In this case we only want the `tombstone` in the JSON if its a tombstone row.  Returning `Optional.empty` means its excluded from the JSON.

Reply:
I think there are better ways of achieving this than using optionals, which seems a little error prone and confusing. For example [`JsonInclude.Include.NON_DEFAULT`](https://fasterxml.github.io/jackson-annotations/javadoc/2.7/com/fasterxml/jackson/annotation/JsonInclude.Include.html#NON_DEFAULT) can be used to omit the serialization of `false`.